File: apps/sdkwork-appstore-flutter-mobile/packages/sdkwork_appstore_flutter_mobile_collection/lib/src/state/collection_state.dart
import '../models/collection_models.dart';

enum CollectionStatus { idle, loading, ready, error }

class CollectionState {
  const CollectionState({
    this.status = CollectionStatus.idle,
    this.lastResult,
    this.error,
  });

  final CollectionStatus status;
  final CollectionPageResult<Object?>? lastResult;
  final String? error;

  CollectionState loading() => const CollectionState(status: CollectionStatus.loading);

  CollectionState loaded(CollectionPageResult<Object?> result) =>
      CollectionState(status: CollectionStatus.ready, lastResult: result);

  CollectionState failed(String message) =>
      CollectionState(status: CollectionStatus.error, error: message);
}

File: apps/sdkwork-appstore-mini-program/packages/sdkwork-appstore-mp-collection/src/state/collectionState.ts
import type { CollectionPageResult } from "../types/collectionModels.js";

export interface CollectionState {
  readonly status: "idle" | "loading" | "ready" | "error";
  readonly lastResult?: CollectionPageResult<unknown>;
  readonly error?: string;
}

export function initialCollectionState(): CollectionState {
  return { status: "idle" };
}

export function reduceCollectionState(
  state: CollectionState,
  event:
    | { type: "load" }
    | { type: "loaded"; result: CollectionPageResult<unknown> }
    | { type: "failed"; error: string },
): CollectionState {
  switch (event.type) {
    case "load":
      return { ...state, status: "loading", error: undefined };
    case "loaded":
      return { status: "ready", lastResult: event.result };
    case "failed":
      return { ...state, status: "error", error: event.error };
  }
}
